### PR TITLE
Default tasks

### PR DESCRIPTION
**Pull request by MaxRobinsonTheGreat**
- Fixed various problems with follow task
- Added avoid task, which returns a higher reward when the actor to avoid is farther away and full reward if the actor to avoid cannot see the agent
- Added functionality for the follow task to give a positive reward or negative reward when the location is reached, as well as an option to make it a terminal task

### CODE REVIEW

File: Source/Holodeck/Tasks/Private/AvoidTask.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	Super::InitializeSensor();
+}
+
+// Allows sensor parameters to be set programmatically from client.
+void UAvoidTask::ParseSensorParms(FString ParmsJson) {
+	Super::ParseSensorParms(ParmsJson);
+
+	TSharedPtr<FJsonObject> JsonParsed;
+	TSharedRef<TJsonReader<TCHAR>> JsonReader = TJsonReaderFactory<TCHAR>::Create(ParmsJson);
+	if (FJsonSerializer::Deserialize(JsonReader, JsonParsed)) {
```
Comment (jaydenmilne) [inline Source/Holodeck/Tasks/Private/AvoidTask.cpp:16]: Should we log an error if we can't deserialize it? `UE_LOG(LogHolodeck, Fatal, TEXT("...` will log the error and crash the engine gracefully

File: Source/Holodeck/Tasks/Private/DistanceTask.cpp [hunk 2]
```diff
@@ -29,22 +24,19 @@ void UDistanceTask::ParseSensorParms(FString ParmsJson) {
...
+			DistanceActorTag = JsonParsed->GetStringField("DistanceActor");
 		}
 
-		if (JsonParsed->HasTypedField<EJson::Array>("DistanceLocation")) {
-			TArray<TSharedPtr<FJsonValue>> LocationArray = JsonParsed->GetArrayField("DistanceLocation");
+		if (JsonParsed->HasTypedField<EJson::String>("GoalActor")) {
+			GoalActorTag = JsonParsed->GetStringField("GoalActor");
+		}
+
+		if (JsonParsed->HasTypedField<EJson::Array>("GoalLocation")) {
```
Comment (jaydenmilne) [inline Source/Holodeck/Tasks/Private/DistanceTask.cpp:34]: Don't forget to update docs & config files https://holodeck.readthedocs.io/en/develop/packages/docs/tasks/distance-task.html

File: Source/Holodeck/Tasks/Private/FollowTask.cpp [hunk 3]
```diff
@@ -20,6 +19,10 @@ void UFollowTask::ParseSensorParms(FString ParmsJson) {
 			ToFollowTag = JsonParsed->GetStringField("ToFollow");
 		}
 
+		if (JsonParsed->HasTypedField<EJson::String>("FollowSocket")) {
```
Comment (jaydenmilne) [inline Source/Holodeck/Tasks/Private/FollowTask.cpp:22]: Don't forget to update docs https://holodeck.readthedocs.io/en/develop/packages/docs/tasks/follow-task.html

Comment (jaydenmilne) [reply Source/Holodeck/Tasks/Private/FollowTask.cpp:22]: If we're going to do this then we should document what sockets are on the agents

File: Source/Holodeck/Tasks/Private/FollowTask.cpp [hunk 4]
```diff
@@ -31,58 +34,57 @@ void UFollowTask::ParseSensorParms(FString ParmsJson) {
 		if (JsonParsed->HasTypedField<EJson::Number>("MinDistance")) {
 			MinDistance = ConvertClientDistanceToUnreal(JsonParsed->GetNumberField("MinDistance"));
 		}
-
-		if (JsonParsed->HasTypedField<EJson::Number>("TargetHeight")) {
```
Comment (jaydenmilne) [inline Source/Holodeck/Tasks/Private/FollowTask.cpp:35]: Don't forget to remove this from the docs

File: Source/Holodeck/Tasks/Private/FollowTask.cpp [hunk 5]
```diff
@@ -31,58 +34,57 @@ void UFollowTask::ParseSensorParms(FString ParmsJson) {
...
-			TargetHeight = ConvertClientDistanceToUnreal(JsonParsed->GetNumberField("TargetHeight"));
-		}
 	}
 }
 
 // Called every frame
 void UFollowTask::TickSensorComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction) {
```
Comment (jaydenmilne) [inline Source/Holodeck/Tasks/Private/FollowTask.cpp:41]: This code seems awfully similar to the new AvoidTask. Is there a way we could generalize it and have less code duplication between the two?

File: Source/Holodeck/Tasks/Private/LocationTask.cpp [hunk 6]
```diff
@@ -15,23 +15,20 @@ void ULocationTask::ParseSensorParms(FString ParmsJson) {
...
-		if (JsonParsed->HasTypedField<EJson::String>("DistanceActor")) {
-			FString ActorName = JsonParsed->GetStringField("DistanceActor");
-			for (TActorIterator<AStaticMeshActor> ActorItr(GetWorld()); ActorItr; ++ActorItr)
-			{
-				if (ActorItr->ActorHasTag(FName(*ActorName))) {
-					DistanceActor = *ActorItr;
-					break;
-				}
-			}
+		if (JsonParsed->HasTypedField<EJson::String>("LocationActor")) {
```
Comment (jaydenmilne) [inline Source/Holodeck/Tasks/Private/LocationTask.cpp:18]: Don't forget to update docs and configs https://holodeck.readthedocs.io/en/develop/packages/docs/tasks/distance-task.html

File: Source/Holodeck/Tasks/Private/LocationTask.cpp [hunk 7]
```diff
@@ -68,3 +79,18 @@ void ULocationTask::TickSensorComponent(float DeltaTime, ELevelTick TickType, FA
 	// Call TaskSensor's Tick to store Reward and Terminal in sensor buffer
 	UTaskSensor::TickSensorComponent(DeltaTime, TickType, ThisTickFunction);
 }
+
+// Get Distance
+float ULocationTask::CalcDistance() {
```
Comment (jaydenmilne) [inline Source/Holodeck/Tasks/Private/LocationTask.cpp:84]: This function seems to be duplicated in three places. Maybe make a helper class?

**Review by jaydenmilne — CHANGES_REQUESTED** [review]

File: Source/Holodeck/Tasks/Private/AvoidTask.cpp (hunk 1 above)
Comment (MaxRobinsonTheGreat) [reply Source/Holodeck/Tasks/Private/AvoidTask.cpp:16]: Added logging

File: Source/Holodeck/Tasks/Private/DistanceTask.cpp (hunk 2 above)
Comment (MaxRobinsonTheGreat) [reply Source/Holodeck/Tasks/Private/DistanceTask.cpp:34]: Updated

File: Source/Holodeck/Tasks/Private/FollowTask.cpp (hunk 4 above)
Comment (MaxRobinsonTheGreat) [reply Source/Holodeck/Tasks/Private/FollowTask.cpp:35]: Removed

File: Source/Holodeck/Tasks/Private/FollowTask.cpp (hunk 5 above)
Comment (MaxRobinsonTheGreat) [reply Source/Holodeck/Tasks/Private/FollowTask.cpp:41]: A few key differences make this code complicated to generalize, but some common functionality has been moved to a helper class

File: Source/Holodeck/Tasks/Private/LocationTask.cpp (hunk 6 above)
Comment (MaxRobinsonTheGreat) [reply Source/Holodeck/Tasks/Private/LocationTask.cpp:18]: updated

File: Source/Holodeck/Tasks/Private/LocationTask.cpp (hunk 7 above)
Comment (MaxRobinsonTheGreat) [reply Source/Holodeck/Tasks/Private/LocationTask.cpp:84]: The actual duplication is the subtraction bit, which I think is simple enough that it doesn't need to be moved to its own function. Getting the start location and end location are actually fairly different between different tasks and shouldn't be generalized in the helper class

**Comment by jaydenmilne** [comment]
BYU-PCCL/holodeck#250

**Review by jaydenmilne — APPROVED** [review]

**Pull request merged**